File: mcode-desktop/src/lib/pairing.ts
export interface GatewayQrPayloadInput {
  name: string
  gatewayProvider: "official" | "custom"
  gatewayBaseUrl?: string
  pairCode: string
  pairSecret: string
}

export interface DesktopGatewayQrPayload {
  version: 2
  name: string
  targetAgent: "mcode-desktop"
  routeMode: "gateway"
  gatewayProvider: "official" | "custom"
  gatewayBaseUrl?: string
  pairCode: string
  pairSecret: string
}

export interface PairOfferInput {
  name: string
  gatewayProvider?: "official" | "custom"
  gatewayBaseUrl?: string
}

export interface PairOfferResult {
  code: string
  secret: string
  qrPayload: string
}

export function buildGatewayQrPayload(input: GatewayQrPayloadInput): DesktopGatewayQrPayload {
  const payload: DesktopGatewayQrPayload = {
    version: 2,
    name: input.name.trim() || "MCode Desktop",
    targetAgent: "mcode-desktop",
    routeMode: "gateway",
    gatewayProvider: input.gatewayProvider,
    pairCode: input.pairCode,
    pairSecret: input.pairSecret,
  }

  const gatewayBaseUrl = input.gatewayBaseUrl?.trim().replace(/\/+$/, "")
  if (gatewayBaseUrl) {
    payload.gatewayBaseUrl = gatewayBaseUrl
  }

  return payload
}

export async function generatePairOffer(input: PairOfferInput): Promise<PairOfferResult> {
  const code = generatePairCode()
  const secret = generateSecret()
  const payload = buildGatewayQrPayload({
    name: input.name,
    gatewayProvider: input.gatewayProvider ?? "official",
    gatewayBaseUrl: input.gatewayBaseUrl,
    pairCode: code,
    pairSecret: secret,
  })

  return {
    code,
    secret,
    qrPayload: JSON.stringify(payload),
  }
}

function generatePairCode(): string {
  const alphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
  const left = randomToken(alphabet, 4)
  const right = randomToken(alphabet, 4)
  return `${left}-${right}`
}

function generateSecret(): string {
  const randomUuid = globalThis.crypto?.randomUUID?.()
  return randomUuid || randomToken("abcdefghijklmnopqrstuvwxyz0123456789", 32)
}

function randomToken(alphabet: string, length: number): string {
  const bytes = new Uint8Array(length)
  globalThis.crypto?.getRandomValues?.(bytes)
  return Array.from(bytes, (value) => alphabet[value % alphabet.length]).join("")
}

File: mcode-desktop/src-tauri/src/pairing.rs
use uuid::Uuid;

use crate::app_state::{GatewayProvider, PairOffer};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct GatewayQrPayloadInput {
    pub name: String,
    pub gateway_provider: GatewayProvider,
    pub gateway_base_url: Option<String>,
    pub pair_code: String,
    pub pair_secret: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PairOfferRequest {
    pub name: String,
    pub gateway_provider: GatewayProvider,
    pub gateway_base_url: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GatewayQrPayload {
    pub version: u8,
    pub name: String,
    pub target_agent: String,
    pub route_mode: String,
    pub gateway_provider: GatewayProvider,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gateway_base_url: Option<String>,
    pub pair_code: String,
    pub pair_secret: String,
}

pub fn build_gateway_qr_payload(input: GatewayQrPayloadInput) -> GatewayQrPayload {
    GatewayQrPayload {
        version: 2,
        name: normalize_name(&input.name),
        target_agent: "mcode-desktop".to_string(),
        route_mode: "gateway".to_string(),
        gateway_provider: input.gateway_provider,
        gateway_base_url: input
            .gateway_base_url
            .map(|value| normalize_url(&value))
            .filter(|value| !value.is_empty()),
        pair_code: input.pair_code,
        pair_secret: input.pair_secret,
    }
}

pub fn generate_pair_offer(input: PairOfferRequest) -> PairOffer {
    let code = generate_pair_code();
    let secret = generate_pair_secret();
    let payload = build_gateway_qr_payload(GatewayQrPayloadInput {
        name: input.name,
        gateway_provider: input.gateway_provider,
        gateway_base_url: input.gateway_base_url,
        pair_code: code.clone(),
        pair_secret: secret.clone(),
    });

    PairOffer {
        code,
        secret,
        qr_payload: serde_json::to_string(&payload).expect("gateway QR payload serializes"),
    }
}

fn generate_pair_code() -> String {
    let raw = Uuid::new_v4().simple().to_string().to_uppercase();
    format!("{}-{}", &raw[0..4], &raw[4..8])
}

fn generate_pair_secret() -> String {
    Uuid::new_v4().simple().to_string()
}

fn normalize_name(value: &str) -> String {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        "MCode Desktop".to_string()
    } else {
        trimmed.to_string()
    }
}

fn normalize_url(value: &str) -> String {
    value.trim().trim_end_matches('/').to_string()
}
